src/index/base.cpp

```diff
@@ -250,8 +259,19 @@ bool BaseIndex::Rewind(const CBlockIndex* current_tip, const CBlockIndex* new_ti
     return true;
 }
 
-void BaseIndex::BlockConnected(const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex)
+void BaseIndex::BlockConnected(ChainstateRole role, const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex)
 {
+    // Ignore events from the assumed-valid chain; we will process its blocks
+    // (sequentially) after it is fully verified by the background chainstate. This
+    // is to avoid any out-of-order indexing.
+    //
+    // TODO at some point we could parameterize whether a particular index can be
```

Comment:
373cf91531b84bfdd06fdf8abf4dca228029ce6b being able to build out of order is necessary for a pruned node.

[note: currently running an IBD on mainnet with `-prune=2000`, `-txindex=1` and `-blockfilterindex=1` to see what happens… I'm guessing it will be missing a range of blocks above the snapshot? If so then we should either refuse to load a snapshot or delay pruning with this combination. If we delay pruning, we should warn the user that enabling the index will temporarily require a lot of extra storage (x GB per month of how old the snapshot is)]

Reply:
I think it syncs ok, but just doesn't prune anything in the snapshot chainstate until the background chainstate has caught up (because of the prune locks introduced in #21726). This should probably be mentioned in some documentation (because users could run into disk space problems with it), see my comment below.

Reply:
I was wondering if those prune locks would kick in. I'll find out.

Reply:
Looks like they did. Documented in #28618.